File: scripts/algorithms/T/Triangle/Triangle.js
var minimumTotal = function(triangle) {
    const memo = {};
    
    function minPath(row, col) {
        let key = `${row}:${col}`;
        
        if (key in memo) {
            return memo[key];
        }
        
        let path = triangle[row][col];
        
        if (row < triangle.length - 1) {
            path += Math.min(minPath(row + 1, col), minPath(row + 1, col + 1));
        }
        
        memo[key] = path;
        
        return path;
    }
    
    return minPath(0, 0);
};

File: scripts/algorithms/T/Triangle/Triangle.java
// Runtime: 6 ms (Top 41.54%) | Memory: 43.9 MB (Top 66.49%)
class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {

        int n = triangle.get( triangle.size() - 1).size();
        int dp[] = new int[n + 1];

        for(int i = triangle.size() - 1; i>=0; i--)
        {
            for(int j = 0; j<triangle.get(i).size(); j++)
                dp[j] = triangle.get(i).get(j) + Math.min(dp[j], dp[j+1]);
        }

        return dp[0];

    }
}